lib/ecto/changeset.ex

```diff
@@ -1203,6 +1205,93 @@ defmodule Ecto.Changeset do
     end
   end
 
+  @doc """
+  Gets the association entry or entries from changes or from the data.
+
+  Returned data is normalized to changesets by default. Pass the `:struct`
+  flag to retrieve the data as structs with changes applied, similar to `get_field/2`.
+
+  ## Examples
+
+      iex> %Author{posts: [%Post{id: 1, title: "hello"}]}
+      ...> |> change()
+      ...> |> get_assoc(:posts)
+      [%Ecto.Changeset{data: %Post{id: 1, title: "hello"}, changes: %{}}]
+
+      iex> %Author{posts: [%Post{id: 1, title: "hello"}]}
+      ...> |> cast(%{posts: [%{id: 1, title: "world"}]}, [])
+      ...> |> cast_assoc(:posts)
+      ...> |> get_assoc(:posts, :changeset)
+      [%Ecto.Changeset{data: %Post{id: 1, title: "hello"}, changes: %{title: "world"}}]
+
+      iex> %Author{posts: [%Post{id: 1, title: "hello"}]}
+      ...> |> cast(%{posts: [%{id: 1, title: "world"}]}, [])
+      ...> |> cast_assoc(:posts)
+      ...> |> get_assoc(:posts, :struct)
+      [%Post{id: 1, title: "world"}]
+
+  """
+  def get_assoc(changeset, name, as \\ :changeset)
+
+  def get_assoc(%Changeset{} = changeset, name, :struct) do
+    get_field(changeset, name)
+  end
+
+  def get_assoc(%Changeset{} = changeset, name, :changeset) do
+    get_relation(:assoc, changeset, name)
+  end
+
+  @doc """
+  Gets the embedded entry or entries from changes or from the data.
+
+  Returned data is normalized to changesets by default. Pass the `:struct`
+  flag to retrieve the data as structs with changes applied, similar to `get_field/2`.
+
+  ## Examples
+
+      iex> %Post{comments: [%Comment{id: 1, body: "hello"}]}
+      ...> |> change()
+      ...> |> get_embed(:comments)
+      [%Ecto.Changeset{data: %Comment{id: 1, body: "hello"}, changes: %{}}]
+
+      iex> %Post{comments: [%Comment{id: 1, body: "hello"}]}
+      ...> |> cast(%{comments: [%{id: 1, body: "world"}]}, [])
+      ...> |> cast_embed(:comments)
+      ...> |> get_embed(:comments, :changeset)
+      [%Ecto.Changeset{data: %Comment{id: 1, body: "hello"}, changes: %{body: "world"}}]
+
+      iex> %Post{comments: [%Comment{id: 1, body: "hello"}]}
+      ...> |> cast(%{comments: [%{id: 1, body: "world"}]}, [])
+      ...> |> cast_embed(:comments)
+      ...> |> get_embed(:comments, :struct)
+      [%Comment{id: 1, body: "world"}]
+
+  """
+  def get_embed(changeset, name, as \\ :changeset)
+
+  def get_embed(%Changeset{} = changeset, name, :struct) do
+    get_field(changeset, name)
+  end
+
+  def get_embed(%Changeset{} = changeset, name, :changeset) do
+    get_relation(:embed, changeset, name)
+  end
+
+  defp get_relation(_tag, %{types: nil}, _name) do
+    raise ArgumentError, "changeset does not have types information"
+  end
+
+  defp get_relation(tag, %{types: types} = changeset, name) do
+    _ = relation!(:get, tag, name, Map.get(types, name))
+    existing = get_change(changeset, name) || get_field(changeset, name)
```

Comment:
```suggestion
    existing = get_field(changeset, name)
```

I feel like I must be missing something something because the part I crossed off is really confusing me: 

- `get_field/3` will already check the changes map
-  if `%{assoc: nil}` is in the change map it will effectively be ignored until `get_field`

Reply:
The issue is that `get_field` will apply the changes, and we don't want that here. But perhaps it is worth just inlining the implementation `get_field` here. It is basically two nested cases anyway.

Reply:
oh I get it. it would be less confusing me to inline but won't complain either way